### PR TITLE
Refactor Serial backend

### PR DESCRIPTION
**Pull request by Rombur**
This PR moves the files associated with the Serial backend to a new `Serial` directory. I have also moved the `ParallelXXX` structs and `UniqueToken` to their own files. The only change in the code is that I've removed some of the macro guards because it's now done at the directory level.

### CODE REVIEW

File: core/src/Serial/Kokkos_Serial_UniqueToken.hpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_CUDA_UNIQUE_TOKEN_HPP
+#define KOKKOS_CUDA_UNIQUE_TOKEN_HPP
```
Comment (ldh4) [inline core/src/Serial/Kokkos_Serial_UniqueToken.hpp:46]: Shouldn't these be `KOKKOS_SERIAL_UNIQUE_TOKEN_HPP`?

File: core/src/CMakeLists.txt [hunk 2]
```diff
@@ -60,9 +60,7 @@ ENDIF()
 
 IF (KOKKOS_ENABLE_SERIAL)
   APPEND_GLOB(KOKKOS_CORE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/Serial/*.hpp)
-ELSE()
-  LIST(REMOVE_ITEM KOKKOS_CORE_SRCS ${CMAKE_CURRENT_SOURCE_DIR}/impl/Kokkos_Serial.cpp)
-  LIST(REMOVE_ITEM KOKKOS_CORE_SRCS ${CMAKE_CURRENT_SOURCE_DIR}/impl/Kokkos_Serial_task.cpp)
+  APPEND_GLOB(KOKKOS_CORE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/Serial/*.cpp)
 ENDIF()
```
[the remark this answers is not in the document]
Comment (Rombur) [reply core/src/CMakeLists.txt:64]: good catch. I'll change it

**Review by JBludau — APPROVED** [review]
looks good

**Comment by Rombur** [comment]
Retest this please

**Review by ldh4 — APPROVED** [review]

**Pull request merged**